qutebrowser/misc/mhtml.py

```diff
@@ -0,0 +1,411 @@
+                    web_frame.findAllElements("img"))
+
+        for element in elements:
+            element_url = element.attribute("src")
+            if not element_url:
+                element_url = element.attribute("href")
+            if not element_url:
+                # Might be a local <script> tag or something else
+                continue
+            absolute_url = web_url.resolved(QUrl(element_url))
+            self.fetch_url(absolute_url)
+
+        styles = web_frame.findAllElements("style")
+        for style in styles:
+            if style.attribute("type", "text/css") != "text/css":
+                continue
+            for element_url in _get_css_imports(style.toPlainText()):
+                element_url = element_url.decode("ascii")
+                self.fetch_url(web_url.resolved(QUrl(element_url)))
+
+        # Search for references in inline styles
+        for element in web_frame.findAllElements("*"):
+            style = element.attribute("style")
+            if not style:
+                continue
+            for element_url in _get_css_imports(style):
+                element_url = element_url.decode("ascii")
+                self.fetch_url(web_url.resolved(QUrl(element_url)))
+
+        # Shortcut if no assets need to be downloaded, otherwise the file would
+        # never be saved. Also might happen if the downloads are fast enough to
+        # complete before connecting their finished signal.
+        self.collect_zombies()
+        if not self.pending_downloads and not self._finished:
+            self.finish_file()
+
+    def fetch_url(self, url):
+        """Download the given url and add the file to the collection.
+
+        Args:
+            url: The file to download as QUrl.
+        """
+        if url.scheme() == "data":
+            return
+        # Prevent loading an asset twice
+        if url in self.loaded_urls:
+            return
+        self.loaded_urls.add(url)
+
+        log.misc.debug("loading asset at %s", url)
+
+        download_manager = objreg.get("download-manager", scope="window",
+                                      window="current")
+        item = download_manager.get(url, fileobj=_NoCloseBytesIO(),
+                                    auto_remove=True)
+        self.pending_downloads.add((url, item))
+        item.finished.connect(
+            functools.partial(self.finished, url, item))
+        item.error.connect(
+            functools.partial(self.error, url, item))
+        item.cancelled.connect(
+            functools.partial(self.error, url, item))
+
+    def finished(self, url, item):
+        """Callback when a single asset is downloaded.
+
+        Args:
+            url: The original url of the asset as QUrl.
+            item: The DownloadItem given by the DownloadManager
+        """
+        self.pending_downloads.remove((url, item))
+        mime = item.raw_headers.get(b"Content-Type", b"")
+        mime = mime.decode("ascii", "ignore")
```

Comment:
I can't be sure of the encoding that the headers use so I chose ASCII while ignoring decode errors - no particular reason, though it would probably be better to set the Content-Type to `None` then, or use some other fallback mechanism...